pkg/controllers/user/monitoring/config_refresh_handler.go

```diff
@@ -0,0 +1,112 @@
+package monitoring
+
+import (
+	"encoding/json"
+	"reflect"
+	"sort"
+
+	"github.com/rancher/rancher/pkg/controllers/user/nslabels"
+	apiv1 "github.com/rancher/types/apis/core/v1"
+	rmonitoringv1 "github.com/rancher/types/apis/monitoring.coreos.com/v1"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/client/monitoring/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	projectNSAnnotation = "project.cattle.io/namespaces"
+)
+
+type ConfigRefreshHandler struct {
+	prometheusClient rmonitoringv1.PrometheusInterface
+	prometheusLister rmonitoringv1.PrometheusLister
+	nsClient         apiv1.NamespaceInterface
+}
+
+func (h *ConfigRefreshHandler) syncNamespace(key string, obj *corev1.Namespace) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return obj, nil
+	}
+	promList, err := h.prometheusLister.List("", labels.NewSelector())
+	if err != nil {
+		return obj, err
+	}
+
+	for _, prometheus := range promList {
+		h.prometheusClient.Controller().Enqueue(prometheus.Namespace, prometheus.Name)
+	}
+
+	return obj, nil
+}
+
+func (h *ConfigRefreshHandler) syncPrometheus(key string, obj *monitoringv1.Prometheus) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return obj, nil
+	}
+	ns, err := h.nsClient.Get(obj.Namespace, metav1.GetOptions{})
+	if err != nil {
+		return obj, err
+	}
+	if len(ns.Labels) == 0 {
+		return obj, nil
+	}
+	projectID, ok := ns.Labels[nslabels.ProjectIDFieldLabel]
+	if !ok {
+		return obj, nil
+	}
+	namespaces, err := h.getProjectNamespaces(projectID)
+	if err != nil {
+		return obj, err
+	}
+
+	annotationNamespaces, err := getAnnotationNamespaces(obj)
+	if err != nil {
+		return obj, err
+	}
+
+	if !reflect.DeepEqual(namespaces, annotationNamespaces) {
```

Comment:
can we safely guarantee the order of these two slices? DeepEquals on lists has bit us mutliple times before. Better to do a set comparison

Reply:
Yes. I sort both slice in `getProjectNamespaces` function and `getAnnotationNamespaces` before they return the slice.